File: src/classes/join/scan_res_item.rs
use std::string::FromUtf8Error;

use crate::consts::{errors::HbrsError, other::Other::KeyLength, response_item::ResponseItem};

pub struct ScanResItem{
    pub ssid: Vec<u8>,
    pub security_type: u8,
    pub encryption_type: u8,
    pub rssi: u8,
}

impl ScanResItem{
    pub fn new(data: Vec<u8>, header_offset: usize) -> Result<Self, HbrsError>{
		if data.len() < ResponseItem::DataSize.get_value(){
			return Err(HbrsError::InvalidScanDataLength);
		}

        let mut ssid = vec![0; KeyLength.get_value()];
        
        ssid.copy_from_slice(&data[header_offset..(header_offset+32)]);

		let security_type = data[header_offset + ResponseItem::SecurityTypePos.get_value()];
        let encryption_type = data[header_offset + ResponseItem::EncryptionTypePos.get_value()];
		let rssi = data[header_offset + ResponseItem::RSSIPos.get_value()];
        
        Ok(Self{
            ssid,
            security_type,
            encryption_type,
            rssi,
        })
    }

    pub fn get_ssid_as_str(&self) -> Result<String, FromUtf8Error>{
        return String::from_utf8(self.ssid.clone());
    }
}

File: lib/classes/join/scan_res_item.py
class ScanResItem:
    def __init__(self, data, header_offset):
        self.ssid = bytearray(33)
        for i in range(33):
            self.ssid[i] = 0

        self.ssid[:32] = data[header_offset:header_offset+32]
        self.ssid = [byte for byte in self.ssid]
        self.security_type = data[header_offset + 33]
        self.encryption_type = data[header_offset + 34]
        self.rssi = data[header_offset + 35]

    def display(self):
        print(self.get_ssid_as_str())
        print(self.security_type)
        print(self.encryption_type)
        print(self.rssi)

    def get_ssid_as_str(self):
        return ''.join(chr(byte) for byte in self.ssid)

    def get_bytes(self):
        return self.ssid + bytes([self.security_type, self.encryption_type, self.rssi])

    def get_ssid(self):
        return self.ssid

    def set_ssid(self, ssid):
        self.ssid = ssid

    def get_security_type(self):
        return self.security_type

    def set_security_type(self, security_type):
        self.security_type = security_type

    def get_encryption_type(self):
        return self.encryption_type

    def set_encryption_type(self, encryption_type):
        self.encryption_type = encryption_type

    def get_rssi(self):
        return self.rssi

    def set_rssi(self, rssi):
        self.rssi = rssi